# Analysis of used car listings on eBay

Analyzing Used Car Listings on eBay Kleinanzeigen

The dataset was originally scraped and uploaded to Kaggle. The version of the dataset we are working with is a sample of 50,000 data points that was prepared by Dataquest including simulating a less-cleaned version of the data.

The data dictionary provided with data is as follows:

dateCrawled - When this ad was first crawled. All field-values are taken from this date.

name - Name of the car.

seller - Whether the seller is private or a dealer.

offerType - The type of listing

price - The price on the ad to sell the car.

abtest - Whether the listing is included in an A/B test.

vehicleType - The vehicle Type.

yearOfRegistration - The year in which which year the car was first registered.

gearbox - The transmission type.

powerPS - The power of the car in PS.

model - The car model name.

kilometer - How many kilometers the car has driven.

monthOfRegistration - The month in which which year the car was first registered.

fuelType - What type of fuel the car uses.

brand - The brand of the car.

notRepairedDamage - If the car has a damage which is not yet repaired.

dateCreated - The date on which the eBay listing was created.

nrOfPictures - The number of pictures in the ad.

postalCode - The postal code for the location of the vehicle.

lastSeenOnline - When the crawler saw this ad last online.


In [1]:
#The aim of this project work is to clean the data and analyze the included used car listings

In [2]:
#As a start I import the python libraries:
import pandas as pd
import numpy as np

In [11]:
#Reading the csv file
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [13]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

In [23]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [25]:
autos.shape
# I observe that the data has 371528 rows and 20 columns

(371528, 20)

# Starting from columns: cleaning

In [17]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

#Change the columns from camelcase to snakecase.

#Change a few wordings to more accurately describe the columns.

In [28]:
autos = autos.rename(columns = {'yearOfRegistration':'registration_year', 'monthOfRegistration':'registration_month', 'notRepairedDamage': 'unrepaired_damage', 'dateCreated': 'ad_created'})

In [29]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'registration_year', 'gearbox', 'powerPS', 'model',
       'kilometer', 'registration_month', 'fuelType', 'brand',
       'unrepaired_damage', 'ad_created', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [40]:
autos = autos.rename(columns = {'dateCrawled':'date_crawled', 'offerType':'offer_type', 'abtest': 'ab_test', 'vehicleType':'vehicle_type', 'powerPS':'power_ps', 'fuelType':'fuel_type', 'nrOfPictures':'nr_of_pictures', 'postalCode':'postal_code', 'lastSeen':'last_seen'})

In [41]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [43]:
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price ab_test vehicle_type  registration_year    gearbox  power_ps  model  \
0    480    test          NaN               1993    manuell         0   golf   
1  18300    test        coupe               2011    manuell       190    NaN   
2   9800    test          suv               2004  automatik       163  grand   
3   1500    test   kleinwagen               2001    manuell        75   golf   
4   3600    test   kleinwagen               2008    manuell        69  fabia   

   kilometer  registration_month fuel_type       brand unrepaired_damage  \
0     150000                   0    benzin  volkswagen               NaN   
1     125000                   5    diesel        audi                ja   
2     125000                   8    diesel        jeep               NaN   
3     150000                   6    benzin  volkswagen              nein   
4      90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

# Exploring the data and further cleaning

In [49]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

       ab_test vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-06 13:45:54  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

From my observations I see that columns:

nr_of_pictures - Look odd and probably needs to be deleted.

registration_month - Min is 0 instead of 1

seller & offer_type - Maybe have the same values

In [58]:
autos['nr_of_pictures'].value_counts()

0    371528
Name: nr_of_pictures, dtype: int64

 The nr_of_pictures column has the value of 0 in every row

In [63]:
autos.drop(['nr_of_pictures'], axis=1, inplace=True)

KeyError: "['nr_of_pictures'] not found in axis"

In [65]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'ab_test',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'postal_code', 'last_seen'],
      dtype='object')

In [67]:
autos['seller'].value_counts()

privat        371525
gewerblich         3
Name: seller, dtype: int64

In [69]:
autos['offer_type'].value_counts()

Angebot    371516
Gesuch         12
Name: offer_type, dtype: int64

In [71]:
autos.drop(['seller', 'offer_type'], axis=1, inplace=True)

In [81]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 17 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   date_crawled        371528 non-null  object
 1   name                371528 non-null  object
 2   price               371528 non-null  int64 
 3   ab_test             371528 non-null  object
 4   vehicle_type        333659 non-null  object
 5   registration_year   371528 non-null  int64 
 6   gearbox             351319 non-null  object
 7   power_ps            371528 non-null  int64 
 8   model               351044 non-null  object
 9   kilometer           371528 non-null  int64 
 10  registration_month  371528 non-null  int64 
 11  fuel_type           338142 non-null  object
 12  brand               371528 non-null  object
 13  unrepaired_damage   299468 non-null  object
 14  ad_created          371528 non-null  object
 15  postal_code         371528 non-null  int64 
 16  la

In [83]:
autos = autos.rename(columns= {'kilometer':'odometer_km'})

In [110]:
autos['odometer_km'] = autos['odometer_km'].astype(int)

In [111]:
autos['price'] = autos['price'].astype(int)

# Exploring the odometer_km and price columns

In [114]:
autos['odometer_km'].describe()

count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

In [108]:
autos['price'].describe()
#There are cars are with the value of 0
#There are cars with the max value of 2.147484e+09 which seems absurd

count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64

In [101]:
autos['price'].value_counts(normalize=True)
#There are 10778/2% of the cars are with the value of 0

0         0.029010
500       0.015261
1500      0.014518
1000      0.012513
1200      0.012365
            ...   
23456     0.000003
171000    0.000003
21830     0.000003
13485     0.000003
8188      0.000003
Name: price, Length: 5597, dtype: float64

In [103]:
autos['price'].value_counts().sort_index(ascending=False).head(20)

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64

In [115]:
autos['price'].value_counts().sort_index(ascending=True).head(50)
#There are numerous cars with different price tags which also need to be adressed/cleaned

0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
21        1
24        1
25       33
26        1
27        1
29        2
30       55
32        1
33        1
35       18
38        1
39        6
40       45
45       16
47        1
49       12
50      327
55       20
58        1
59        6
60       58
65       19
66        2
69        3
70       69
74        1
75       54
77        1
79        4
80      146
Name: price, dtype: int64

Given that ebay has auction option in the website, I will remove the listings from 1-50 and everything above 250,000$.

Therefore, the remaining listings will be from 50 up to 250,000$ . 

In [118]:
autos = autos[autos['price'].between(50, 250000)]

In [121]:
autos['price'].describe()

count    358958.000000
mean       5904.827150
std        8525.234657
min          50.000000
25%        1250.000000
50%        3053.000000
75%        7499.000000
max      250000.000000
Name: price, dtype: float64

# Analyzing the date columns

The number of columns with date information:

date_crawled

registration_month

registration_year

ad_created

last_seen

These are a combination of dates that were crawled, and dates with meta-information from the crawler. 
The non-registration dates are stored as strings.

I explore each of these columns to learn more about the listings.

In [151]:
autos[['date_crawled', 'registration_month', 'registration_year', 'ad_created', 'last_seen']].head(10)

date_crawled  registration_month    registration_year  \
0   2016-03-24 11:52:17                   0  2016-03-24 11:52:17   
1   2016-03-24 10:58:45                   5  2016-03-24 10:58:45   
2   2016-03-14 12:52:21                   8  2016-03-14 12:52:21   
3   2016-03-17 16:54:04                   6  2016-03-17 16:54:04   
4   2016-03-31 17:25:20                   7  2016-03-31 17:25:20   
5   2016-04-04 17:36:23                  10  2016-04-04 17:36:23   
6   2016-04-01 20:48:51                   8  2016-04-01 20:48:51   
8   2016-04-04 23:42:13                   8  2016-04-04 23:42:13   
9   2016-03-17 10:53:50                   0  2016-03-17 10:53:50   
10  2016-03-26 19:54:18                  12  2016-03-26 19:54:18   

             ad_created            last_seen  
0   2016-03-24 00:00:00  2016-04-07 03:16:57  
1   2016-03-24 00:00:00  2016-04-07 01:46:50  
2   2016-03-14 00:00:00  2016-04-05 12:47:46  
3   2016-03-17 00:00:00  2016-03-17 17:40:17  
4   2016-03-31 00:00:00  2016-04-06 10:17:21  
5   2016-04-04 00:00:00  2016-04-06 19:17:07  
6   2016-04-01 00:00:00  2016-04-05 18:18:39  
8   2016-04-04 00:00:00  2016-04-04 23:42:13  
9   2016-03-17 00:00:00  2016-03-31 17:17:06  
10  2016-03-26 00:00:00  2016-04-06 10:45:34

In [152]:
(autos['date_crawled'] 
.str[:10]
.value_counts(normalize = True, dropna= False)
.sort_index()
)

2016-03-05    0.025557
2016-03-06    0.014464
2016-03-07    0.035642
2016-03-08    0.033466
2016-03-09    0.034149
2016-03-10    0.032639
2016-03-11    0.032753
2016-03-12    0.036300
2016-03-13    0.015790
2016-03-14    0.036347
2016-03-15    0.033433
2016-03-16    0.030199
2016-03-17    0.031650
2016-03-18    0.013107
2016-03-19    0.035258
2016-03-20    0.036369
2016-03-21    0.035653
2016-03-22    0.032466
2016-03-23    0.031987
2016-03-24    0.029898
2016-03-25    0.032759
2016-03-26    0.031987
2016-03-27    0.030196
2016-03-28    0.035054
2016-03-29    0.034127
2016-03-30    0.033564
2016-03-31    0.031876
2016-04-01    0.034177
2016-04-02    0.035129
2016-04-03    0.038860
2016-04-04    0.037665
2016-04-05    0.012751
2016-04-06    0.003126
2016-04-07    0.001605
Name: date_crawled, dtype: float64

In [153]:
(autos['date_crawled'] 
.str[:10]
.value_counts(normalize = True, dropna= False)
.sort_values()
)

2016-04-07    0.001605
2016-04-06    0.003126
2016-04-05    0.012751
2016-03-18    0.013107
2016-03-06    0.014464
2016-03-13    0.015790
2016-03-05    0.025557
2016-03-24    0.029898
2016-03-27    0.030196
2016-03-16    0.030199
2016-03-17    0.031650
2016-03-31    0.031876
2016-03-23    0.031987
2016-03-26    0.031987
2016-03-22    0.032466
2016-03-10    0.032639
2016-03-11    0.032753
2016-03-25    0.032759
2016-03-15    0.033433
2016-03-08    0.033466
2016-03-30    0.033564
2016-03-29    0.034127
2016-03-09    0.034149
2016-04-01    0.034177
2016-03-28    0.035054
2016-04-02    0.035129
2016-03-19    0.035258
2016-03-07    0.035642
2016-03-21    0.035653
2016-03-12    0.036300
2016-03-14    0.036347
2016-03-20    0.036369
2016-04-04    0.037665
2016-04-03    0.038860
Name: date_crawled, dtype: float64

From my observations, it looks like the data crawled was roughly 1 month worth of data. More particluarly data on the month : 
March 

In [185]:
autos['registration_year'].astype(float)

ValueError: could not convert string to float: '2016-03-24 11:52:17'

From the observation I see that the year that the first car was registered in year 1000. Also, another year that is odd is 9999.

In [157]:
autos = autos[autos['registration_year'].between(1886, 2018)]

TypeError: '>=' not supported between instances of 'str' and 'int'

In [169]:
autos['registration_year'].astype(int)

ValueError: invalid literal for int() with base 10: '2016-03-24 11:52:17'

In [171]:
autos['registration_month'].describe()

count    358958.000000
mean          5.809131
std           3.678888
min           0.000000
25%           3.000000
50%           6.000000
75%           9.000000
max          12.000000
Name: registration_month, dtype: float64